### ISSUE TITLE
Resuming snapshot from a Nestrov solver gets initialized with SGDSolver and trains from cratch

### ISSUE THREAD

**Issue by Coderx7** [BVLC/caffe#5581]
### Issue summary
When I try to resume from a snapshot which uses "`Nesterov`", it instead initializes the network using a `SGDSolver`, and thus no resuming takes place! everything starts from scratch!
For example, When I took a snapshot at 45K, my top1 accuracy was lets say 25 and loss was around 3.6
when I resumed from that iteration (45K), the loss was 87 and accuracy was 0.001.
```
I0501 00:19:56.461567 16032 solver.cpp:330] Iteration 45000, Testing net (#0)
I0501 00:19:56.461567 16032 net.cpp:676] Ignoring source layer accuracy_top-1-train
I0501 00:19:56.461567 16032 net.cpp:676] Ignoring source layer accuracy_top-5-train
I0501 00:20:38.880079 16032 blocking_queue.cpp:49] Waiting for data
I0501 00:22:49.227219 11012 data_layer.cpp:73] Restarting data prefetching from start.
I0501 00:22:49.297024 16032 solver.cpp:397]     Test net output #0: accuracy_top-1 = 0.25042
I0501 00:22:49.297024 16032 solver.cpp:397]     Test net output #1: accuracy_top-5 = 0.50046
I0501 00:22:49.297024 16032 solver.cpp:397]     Test net output #2: loss = 3.62476 (* 1 = 3.62476 loss)
```
 and when resumed : 
```
I0501 01:10:59.971001 10680 net.cpp:255] Network initialization done.
I0501 01:10:59.972002 10680 solver.cpp:56] Solver scaffolding done.
I0501 01:10:59.974001 10680 caffe.cpp:243] Resuming from examples/net/snapshots/slim_iter_45000.solverstate
I0501 01:11:00.294180 10680 sgd_solver.cpp:318] SGDSolver: restoring history
I0501 01:11:00.295181 10680 caffe.cpp:249] Starting Optimization
I0501 01:11:00.295181 10680 solver.cpp:272] Solving SlimNet_300K_NoDrp
I0501 01:11:00.295181 10680 solver.cpp:273] Learning Rate Policy: multistep
I0501 01:11:00.297184 10680 solver.cpp:330] Iteration 45000, Testing net (#0)
I0501 01:11:00.299199 10680 net.cpp:676] Ignoring source layer accuracy_top-1-train
I0501 01:11:00.299199 10680 net.cpp:676] Ignoring source layer accuracy_top-5-train
I0501 01:11:00.568996 10680 blocking_queue.cpp:49] Waiting for data
I0501 01:14:40.206624 15004 data_layer.cpp:73] Restarting data prefetching from start.
I0501 01:14:40.250162 10680 solver.cpp:397]     Test net output #0: accuracy_top-1 = 0.001
I0501 01:14:40.250162 10680 solver.cpp:397]     Test net output #1: accuracy_top-5 = 0.00508
I0501 01:14:40.250162 10680 solver.cpp:397]     Test net output #2: loss = 87.2489 (* 1 = 87.2489 loss)
I0501 01:14:40.728931 10680 solver.cpp:218] Iteration 45000 (-nan iter/s, 220.427s/100 iters), loss = 7.10973
I0501 01:14:40.728931 10680 solver.cpp:237]     Train net output #0: accuracy_top-1-train = 0
I0501 01:14:40.728931 10680 solver.cpp:237]     Train net output #1: accuracy_top-5-train = 0.00390625
I0501 01:14:40.728931 10680 solver.cpp:237]     Train net output #2: loss = 7.10973 (* 1 = 7.10973 loss)
I0501 01:14:40.728931 10680 sgd_solver.cpp:105] Iteration 45000, lr = 0.02
```

### Steps to reproduce
Simply define `Nesterov `in a solver, train for some time and then try to resume it.  This is the config I use in my solver.
```
base_lr: 0.02
display: 100
max_iter: 1000000
lr_policy: "multistep"
gamma: 0.1
momentum: 0.9
weight_decay: 0.0001
snapshot: 5000
solver_mode: GPU
random_seed: 1705
stepvalue: 50000
stepvalue: 61000
type: "Nesterov"
```

### Your system configuration
Operating system: Windows 10 x64
Compiler:Visual Studio 2015
CUDA version (if applicable): 8
CUDNN version (if applicable):5.1
BLAS:
Python or MATLAB version (for pycaffe and matcaffe respectively):python 3.5

**Comment by willyd** [comment]
@Coderx7 what interface are you using? command line `caffe.exe`, python, MATLAB?

**Comment by Coderx7** [comment]
@willyd : I'm using Command line (caffe.exe)

**Comment by willyd** [comment]
Although the log is misleading there is no issue here. Since NesterovSolver is a subclass of SGDSolver they share some code. It happens that the log has a hard coded reference to SGDSolver.

The log comes from:
https://github.com/BVLC/caffe/blob/master/src/caffe/solvers/sgd_solver.cpp#L318

If we replace this line by:
```
LOG(INFO) << this->type() << "Solver: restoring history";
```
you will see NesterovSolver instead of SGDSolver in the log.

**Comment by Coderx7** [comment]
This is not the case, how do you explaining the change in loss? resuming is supposed to resume the training where ever it was left off! 
the loss after resuming becomes 87! instead of being just 3! and continues to decrease as if there were never any previously saved states! 
Whats the point of resuming here then?

**Comment by willyd** [comment]
Right. I totally missed that part of the issue. Can you reproduce this with a simple example like MNIST?

**Comment by Coderx7** [comment]
@willyd : I couldn't replicate the same issue with MNIST, however on my imagenet this kept happening. 
I used another snapshot and it started working. It seems the snapshot I was using was somehow corrupted and caused this weird behavior . 
Thank you for your assistance
